How to use Cabal in Windows

jasagredo opened the pull request:
This is my humble attempt at providing some documentation for how to use GHC and Cabal in Windows. It aims to resolve #9713 , however all I wrote here is based on my experience and might not be The Correct Way To Do Things :tm:. I could use some thorough reviews.

endgame commented on doc/how-to-run-in-windows.rst:
```diff
@@ -0,0 +1,352 @@
+- On PowerShell:
+  ::
+
+     PS> $env:Path += ';C:\msys64\usr\bin;C:\msys64\ucrt64\bin'
```
Do we need these `PATH` fixes, or will they be taken care of by earlier config suggestions? I am starting to become confused.

bradrn replied:
These `PATH` fixes are completely unnecessary and not recommended: see @hasufell’s comments.

jasagredo replied:
Removed, thanks

endgame commented on doc/how-to-run-in-windows.rst:
```diff
@@ -0,0 +1,352 @@
+Once these directories are in the path, any UNIX-like utilities we might need
+will be in scope, so configuring packages should be able to find them.
+
+Using Cabal on Windows
```
Should this section swap with the one immediately above? I feel like `cabal` is the standard way most people are going to trigger GHC.

endgame commented on doc/how-to-run-in-windows.rst:
```diff
@@ -0,0 +1,352 @@
+normally cabal will be able to invoke GHC as said location will be in the
+``PATH``.
+
+All the above extra configurations will also be needed when using ``cabal``, and
+one can provide all of them via the following local project file:
+
+::
+
+   msys2> cat << EOF > cabal.project.local
+   package *
+     extra-include-dirs: C:\msys64\ucrt64\include
+     extra-lib-dirs: C:\msys64\ucrt64\lib
+     extra-prog-path: C:\msys64\ucrt64\bin,
+                      C:\msys64\usr\bin
+   EOF
```
I read this section and think "ok, this sounds like it's going to solve my problem" and then immediately find out that it's not the best way to solve my problem, because the guide then talks about setting up global configuration. I'd consider putting this after the best solution or omitting it.

endgame commented on doc/how-to-run-in-windows.rst:
```diff
@@ -0,0 +1,352 @@
+
+::
+
+   > cat C:\\Users\\<YourName>\\AppData\\Roaming\\cabal\\config
+   ...
+   extra-include-dirs: C:\msys64\ucrt64\include
+   ...
+   extra-lib-dirs: C:\msys64\ucrt64\lib
+   ...
+   extra-prog-path: C:\ghcup\bin,
+                    C:\Users\<YourName>\AppData\Roaming\cabal\bin,
+                    C:\msys64\ucrt64\bin,
+                    C:\msys64\usr\bin
+   ...
```
Why isn't ghcup getting this right for us?

hasufell replied:
It is. But this shouldn't be part of the cabal documentation.

jasagredo replied:
GHCup is setting it for us. I hope I made it clear in the new version.

If this should not be part of the cabal documentation, then what should be part of it? I don't understand your point.

endgame reviewed (commented):
I'm not on the Cabal team and I haven't used Windows for years, but this feels important enough that I didn't want it to sit unreviewed. It's great that you wrote this, and I think with a bit of focusing on the intended reader, it'll be a very useful addition to the manual.

Overall, I think it's important to keep the main reader of this document in mind: someone who has just installed Haskell (probably through GHCup) and needs to get cabal working. They will be very frustrated if things appear to work but then mysterious msys failures cause breakage the first time they build a package that wants to run a `./configure` script or do unixy things.

I think information which supports that goal should come first, and each section can drill into the details as much as is necessary. I get the sense that this PR contains a lot of hard-won knowledge; while it's good to have those details written down somewhere, I believe the main reader of a document like this will want to just get cabal going (and may drill down if things aren't working right).

hasufell commented on doc/how-to-run-in-windows.rst:
```diff
@@ -0,0 +1,352 @@
+
+MSYS2 comes with different environments. You can learn more about the different
+environments in the `MSYS2 documentation
+<https://www.msys2.org/docs/environments/>`_. We will be using ``UCRT64`` in the
```
We should be using `MINGW64` in the examples, since this is the default environment for ghcup and stack.

bradrn replied:
I thought you explicitly recommended UCRT64 the other day? Though I can’t find the post, so I may be misremembering.

jasagredo replied:
In the new version I mention explicitly that GHCup uses MINGW64 by default. To my understanding, MSYS2 already explicitly stated that the default should be UCRT64, that's why I use it and why I advertised it here, but indeed, the default is MINGW64. We will have to update the documentation once GHCup decides to move to UCRT64.

hasufell commented on doc/how-to-run-in-windows.rst:
```diff
@@ -0,0 +1,352 @@
+<https://www.msys2.org/docs/environments/>`_. We will be using ``UCRT64`` in the
+examples below but it can be swapped by ``CLANG64`` (as long as the external
+packages you want to use are available for the ``CLANG64`` environment, see
+:ref:`Using system libraries with GHC`).
```
Basically just by changing `extra-include-dirs`, `extra-lib-dirs` and `extra-prog-path`. Cabal probably shouldn't cross-link to ghcup (at least that's my opinion).

bradrn replied:
Running `mintty` should pop up a dialog allowing the user to choose an environment. I’m a bit surprised this makes no mention of `mintty`, actually — it’s the easiest way to start up the  MSYS2 environment bundled with GHCup and Stack.

jasagredo replied:
I suspect you were asking "why should I choose one environment or other". MSYS2 changed the default to `ucrt`-based environments, ([see](https://www.msys2.org/news/#2022-10-29-changing-the-default-environment-from-mingw64-to-ucrt64)), however GHCup and Stack still use `MINGW64`. Also see the page linked there: https://www.msys2.org/docs/environments/. It describes the difference between gcc and clang-based as well as msvcrt and ucrt based environments. It also says: `If you are unsure, go with UCRT64.`

If you were asking "what steps do I need to take to change environments" then Julian or Brad's answers are correct.

mpilgrem replied:
By way of update about Stack's approach: the `master` branch version allows the MSYS2 environment to be configured: https://github.com/commercialhaskell/stack/issues/6457.

hasufell commented on doc/how-to-run-in-windows.rst:
```diff
@@ -0,0 +1,352 @@
+<https://www.haskell.org/ghcup/>`_. There are however two different ways of
+using GHCup on Windows:
+
+GHCup without having installed MSYS2 already
```
I recommend not  to provide ghcup specific documentation. Leave that to ghcup developers.

jasagredo replied:
This section is gone. Now I just refer to GHCup for instructions.

hasufell commented on doc/how-to-run-in-windows.rst:
```diff
@@ -0,0 +1,352 @@
+
+::
+
+   PS> Set-ExecutionPolicy Bypass -Scope Process -Force;[System.Net.ServicePointManager]::SecurityProtocol = [System.Net.ServicePointManager]::SecurityProtocol -bor 3072; try { Invoke-Command -ScriptBlock ([ScriptBlock]::Create((Invoke-WebRequest https://www.haskell.org/ghcup/sh/bootstrap-haskell.ps1 -UseBasicParsing))) -ArgumentList $true } catch { Write-Error $_ }
```
As above... I don't think cabal should provide this type of information. It is likely going to be out of date and cause issues when I change the default arguments etc.

jasagredo replied:
Agree. This section is now gone.

Kleidukos commented:
@bradrn would you mind taking a look?

hasufell commented on doc/how-to-run-in-windows.rst:
```diff
@@ -0,0 +1,352 @@
+    prompt and when in a native Windows shell we will use the ``PS>`` or ``cmd>``
+    prompt. If either shell will do, we will just use ``>``.
+
+A small note on configuring and updating MSYS2
```
I find the order here confusing. We start with notes about updating msys2, but there was no information on how to install it.

jasagredo replied:
I would have expected the user to click on the links above this part, read a bit about what is MinGW and MSYS2 and install MSYS2. It is true that I don't say that explicitly. Anyways, this section is gone now.

bradrn commented:
Thanks for the reminder! I’ll read through it later today.

hasufell commented on doc/how-to-run-in-windows.rst:
```diff
@@ -0,0 +1,352 @@
+As said above, the binaries that we might need to use are (by default) located
+in two places: ``C:\msys64\usr\bin`` and ``C:\msys64\<environment>\bin``. The
+easiest way to give GHC access to these executables is by including those
+directories in your ``PATH``.
```
Again: no, not recommended.

jasagredo replied:
To me "include in your PATH" does not necessarily mean "set your PATH variable to include this". But I see this is causing a lot of confusion so I left this out.

bradrn commented on doc/how-to-run-in-windows.rst:
```diff
@@ -0,0 +1,352 @@
+A small note on configuring and updating MSYS2
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Before starting with MSYS2, one should make sure it is updated and properly configured. For that:
```
The first four commands are most important and unlikely to fall out of date. The others seem more opinionated to me — I don’t recall ever having needed to run them, though they might have been useful if I’d known.

jasagredo replied:
These instructions are now gone. They were here just to provide some context for new users.

hasufell commented on doc/how-to-run-in-windows.rst:
```diff
@@ -0,0 +1,352 @@
+
+   msys2> cat << EOF > cabal.project.local
+   package *
+     extra-include-dirs: C:\msys64\ucrt64\include
+     extra-lib-dirs: C:\msys64\ucrt64\lib
+     extra-prog-path: C:\msys64\ucrt64\bin,
+                      C:\msys64\usr\bin
```
This is the **main information for dealing with msys2 on cabal** and is way too late in this documentation.

Additionally, I don't think this generally belongs in `cabal.project`, but in `cabal.config`.

bradrn replied:
`cabal.config` strikes me as the right place for this too. These instructions aren’t even correct if MSYS2 has been installed via GHCup or Stack — they are computer-specific.

bradrn replied:
Actually, on reflection… perhaps `cabal.project.local` is the right location for this after all? The Cabal user guide already recommends `cabal.project.local` for local and non–version-controlled changes.

runeksvendsen replied:
@bradrn `cabal.project*` is project-specific, but these paths seems to me like they should be global

bradrn commented on doc/how-to-run-in-windows.rst:
```diff
@@ -0,0 +1,352 @@
+
+::
+
+   # Update the system, run twice
```
I remember the ‘run twice’ thing from when I was on Windows. As I recall, the first command actually halts midway through with explicit instructions to do this, along the lines of ‘close the terminal without Ctrl+C and then re-run the command’.

runeksvendsen replied:
FWIW: the GHCup installer does does it twice as well (https://github.com/haskell/ghcup-hs/blob/2fdf896fbdd338ba6b4244c47b10dcd4067552e0/scripts/bootstrap/bootstrap-haskell.ps1#L497-L501) and I found this README with the same instructions (but without a reason): https://github.com/xournalpp/xournalpp/blob/master/readme/WindowsBuild.md#update-msys2

jasagredo replied:
It is something that usually happens also on ArchLinux I think, that system upgrades need to be run twice. I don't know the reason though.

hasufell reviewed (commented):
In general, I'm mildly skeptical of the current version of this PR.

I do not like the idea that cabal documentation provides extensive information about how GHCup does things, how to install it and run it. That is in the GHCup documenation. Let's please keep things separate.

In addition, I find this is quite verbose and the main information (adding things to `extra-include-dirs`, `extra-lib-dirs` and `extra-prog-path`) is way too far down.

The information about adjusting PATH I consider incorrect.

hasufell commented on doc/how-to-run-in-windows.rst:
```diff
@@ -0,0 +1,352 @@
+  These are referred to as belonging to the ``msys`` repository.
+
+  In order to use tools from ``C:\msys64\usr\bin`` we need to make sure it is
+  present in the ``PATH``.
```
(github swallowed this comment, so here it is again)

This is not recommended, because changing PATH affects the dynamic link library search path for all other tooling as well: https://learn.microsoft.com/en-us/windows/win32/dlls/dynamic-link-library-search-order#standard-search-order-for-unpackaged-apps

Hence, we should rely on `extra-prog-path` and invoke things via cabal or in a proper msys2 shell.

bradrn replied:
Completely agreed with this. There are a lot of resources out there which recommend changing PATH, and I’ve never understood why, when running through the MSYS2 shell is so much easier and more robust. (See also my mention of `mintty` above.)

jasagredo replied:
Well there is a discrepancy here. What if I want to compile something on PowerShell and not on an MSYS2 shell? Isn't one of the points of GHC bundling a MINGW toolchain allowing the user to call it from PowerShell and not rely on a system's MSYS2? What if I need to call `make` from PowerShell?

I just presented what needs to happen, those directories need to be in the PATH, either by setting the env var or by telling cabal to dynamically extend the path with them when calling other tools. Setting the env var indeed will impact other things.

Anyways, this section is now gone. In the cabal documentation I now just mention how to run Cabal, which means that there only a need to modify `extra-prog-path` and nothing else.

bradrn replied:
> What if I want to compile something on PowerShell and not on an MSYS2 shell?

Then, if really necessary, you can change the PATH temporarily. But I’ve almost never found it necessary — firing up one of the shells installed with MSYS2 is much more convenient.

hasufell replied:
> What if I need to call make from PowerShell?

`ghcup run -m -- make`.

But how is that a common use case?

jasagredo replied:
Admittedly not that common, but I was just trying to make sure the UX was smoother by _suggesting_ the PATH modification, for a seamlessly integration in PS.

jasagredo replied:
Only after this PR https://github.com/haskell/cabal/pull/9527

bradrn commented on doc/how-to-run-in-windows.rst:
```diff
@@ -0,0 +1,352 @@
+     "icon": "C:\\msys64\\ucrt64.ico",
+     "name": "MSYS2 / UCRT64",
+     "startingDirectory": "C:/Users/<YourName>"
+   }
```
This is entirely unnecessary. Cabal documentation does not need to advise the user on which terminal configuration to use, let alone their choice of colour scheme!

jasagredo replied:
This was more of a recommendation, thats why it says "I use this" and not "this is the profile that you should use". I was hesitant to add this part but I thought it might be kind for a new user that doesn't really get how the msys2_shell has to be invoked to have an already crafted example.

Anyways, this part is also gone now, I do not mention any terminals. The user can use whatever they want (they also could do it in this version of the document, these were just recommendations).

bradrn commented on doc/how-to-run-in-windows.rst:
```diff
@@ -0,0 +1,352 @@
+For more information about other environment variables, refer to `the GHCup
+documentation <https://www.haskell.org/ghcup/guide/#env-variables>`_.
+
+It is advisable also to set ``GHCUP_MSYS2=C:\msys64`` as a System variable on
+Windows:
```
I’d prefer that phrasing, but it probably doesn’t matter.

In any case, this is GHCup-specific, and I don’t believe it belongs in Cabal documentation.

jasagredo replied:
Removed from the documentation.

bradrn commented on doc/how-to-run-in-windows.rst:
```diff
@@ -0,0 +1,352 @@
+<https://packages.msys2.org/base/mingw-w64-openssl>`_ (and in particular for
+CLANG64 or UCRT64). When installing packages from MSYS2, one should pick the
+options that show ``ucrt64/`` or ``clang64/`` as prefixes, and avoid the ones
+that show ``msys/``.
```
MSYS libraries rely on Cygwin, which is undesirable in general.

runeksvendsen replied:
Note: the default for both Stack and GHCup is MINGW64. GHC itself transitioned from MINGW64 to CLANG64 about two years ago: https://gitlab.haskell.org/ghc/ghc/-/issues/21019

jasagredo replied:
It was explained above:
```
We might
  use tools that are in ``C:\msys64\usr\bin`` but we should not link against
  libraries in ``C:\msys64\usr\lib`` as those target ``msys-2.0.dll`` as a
  runtime dependency which is a fork of Cygwin and not a native Windows DLL.
  These are referred to as belonging to the ``msys`` repository.
```

bradrn reviewed (commented):
Just read through this document. In general, I agree with @hasufell: this is overly long, poorly structured, covers many extraneous details, and contains some recommendations which may be actively harmful. To me, it reads more like ‘my recommendations for using Haskell on Windows’, rather than something which can be included in Cabal documentation.

However, I think it’s salvageable. I may try to condense this into a shorter and more focussed guide, then make my own PR.

EDIT: To make myself clear, I don’t think this is a *bad* document necessarily. It would make a fine standalone article. It’s just that it’s wider in scope than is necessary for Cabal-specific documentation.

jasagredo commented:
I tried to build up the information step by step, first with GHC alone then with cabal. I think that although it is at the end of the document, a user first checks the list of subsections, at least that is what I would do. 

I kind of prefer to be extra verbose, explain too many things but still have things explained somewhere. I'm unaware of a resource that tells you how to use Haskell on Windows, the closest thing is GHC wiki pages which are either outdated or talk only about how to build GHC alone on windows.

However if one doesn't feel that way, I can see how this can be seen as too verbose and overly specific for the Cabal documentation. I will trim it down. 

In terms of Documentation relevant to Cabal, I'm unsure then what to mention. I propose the following:

1. Install via ghcup. Refer to its documentation.
2. Libraries need to be visible in extra-lib-dirs. Includes need to be visible in extra-include-dirs. Unix tools need to be visible in extra-prog-path. Ghcup takes care of it for you.
3. End

I will try to answer why I chose to write the chunks where questions were asked above.

jasagredo commented:
@bradrn @hasufell @endgame First of all, kindly thank you for your reviews. 

I have pushed another commit that fully re-does the document, this time limiting myself to mention the things that seem relevant to "what do I need to do to run Cabal on windows". I omitted side explanations on why are things as they are or giving context on what is being used underneath and why.

I hope this looks more suitable for the Cabal documentation.

jasagredo commented on doc/how-to-run-in-windows.rst:
```diff
@@ -0,0 +1,352 @@
+
+    When the code snippets must be run in a MSYS2 shell, we will use the ``msys2>``
+    prompt and when in a native Windows shell we will use the ``PS>`` or ``cmd>``
+    prompt. If either shell will do, we will just use ``>``.
```
Yeah, it would have been possible. I prefer to be explicit. In any case this is gone now.

jasagredo commented on doc/how-to-run-in-windows.rst:
```diff
@@ -0,0 +1,352 @@
+This built-in toolchain allows GHC to run on native Windows by using the bundled
+``clang`` compiler.
+
+Once GHCup installed GHC, it is possible to invoke just it from any Windows
```
Much clearer reword indeed, but this line is gone.

bradrn commented:
As it happens, I’ve been working on [my own version](https://github.com/bradrn/cabal/blob/how-to-windows/doc/how-to-run-in-windows.rst) of this document. I’m no longer going to submit it as its own PR, since I believe @jasagredo’s new version is superior. However, if there’s anything I’ve written there which looks useful to integrate into this PR, feel free to copy it over!

jasagredo commented:
@bradrn I would not mention that it is _strongly recommended_ to use the installations that come with GHCup and Stack. If a user *only* wants to do Haskell on their machine then maybe, but what if you want to install an editor such as Emacs by compiling it in MSYS2? Or do work in some other language? Must I be using the GHCup MSYS2? I don't think so.

bradrn commented:
> @bradrn I would not mention that it is _strongly recommended_ to use the installations that come with GHCup and Stack.

I meant, of course, for Haskell-specific work. I assumed that was clear from context.

What I’d like to avoid is the situation where someone installs GHCup, then sees that it can integrate with MSYS2 so they go and install MSYS2 separately. The end result is that they’ve mistakenly ended up with an MSYS2 installation which GHCup *doesn’t* know about, which could get confusing.

jasagredo commented:
Indeed. That is what I tried to explain before by saying that GHCup can work with an existing MSYS2 and trying to outline how to do so. 

However it seems this is out of the scope for the Cabal documentation. I might file a ticket to GHCup to mention this then.

bradrn commented:
> However it seems this is out of the scope for the Cabal documentation. I might file a ticket to GHCup to mention this then.

This would be a good idea. I think it’s at least fair to mention in the Cabal documentation that Cabal requires knowing the location of MSYS2, and describe some places where it might be found.

I mean, for that matter, your latest version strongly recommends using GHCup in the first place. If that’s acceptable, then I think my recommendations certainly should be too.

jasagredo commented:
> Cabal requires knowing the location of MSYS2

This is only true if you happen to need something from MSYS2. If you are only compiling packages which do not rely on external tools then you don't need an MSYS2. So Cabal doesn't _require_ knowing this. In fact it is the User the one that _wants_ cabal to know these.

> your latest version strongly recommends using GHCup in the first place

It recommends using GHCup to install a Haskell environment. I chose deliberately to defer the question about "what if I already have an MSYS2 installation?" to GHCup's documentation where it seems this should be mentioned. I tried to explain it here but this was widely rejected. To be clear I'm *not* saying that I recommend using GHCup's MSYS2. I might need to make this clear in the document but I don't know how to do so without stepping over what is said to be GHCup's business.

bradrn commented:
> > Cabal requires knowing the location of MSYS2
> 
> This is only true if you happen to need something from MSYS2. If you are only compiling packages which do not rely on external tools then you don't need an MSYS2. So Cabal doesn't _require_ knowing this. In fact it is the User the one that _wants_ cabal to know these.

Aargh, imprecise wording again! Of course, this whole conversation is only relevant if you need to use external packages.

> > your latest version strongly recommends using GHCup in the first place
> 
> It recommends using GHCup to install a Haskell environment.

Yes, that’s what I was referring to. To me this seems a far stronger (and less relevant) statement than ‘use GHCup/Stack’s MSYS2 environment, if you’re using GHCup/Stack’.

jasagredo commented:
I have pushed a new version. This time changing the title and moving the line `GHCup will install its own MSYS2 system in your computer unless told not to do so, refer to `its documentation <https://www.haskell.org/ghcup/install/#windows_1>`_ for more information.` up.

hasufell commented on doc/how-to-run-in-windows.rst:
```diff
@@ -0,0 +1,142 @@
+Ensure that Cabal can call UNIX-like tools
+------------------------------------------
+
+As Cabal needs sometimes to call UNIX-like tools that come with MSYS2 (such as
```
```suggestion
As Cabal sometimes needs to call UNIX-like tools that come with MSYS2 (such as

```

jasagredo replied:
Done, thanks!

hasufell commented on doc/how-to-run-in-windows.rst:
```diff
@@ -0,0 +1,142 @@
+
+As Cabal needs sometimes to call UNIX-like tools that come with MSYS2 (such as
+``make`` or even ``git``), the directories where those are located need to be
+made visible in the ``PATH``. For that, Cabal provides the ``extra-prog-path``
```
```suggestion
made visible in the ``PATH`` when executing ``cabal``. For that, Cabal provides the ``extra-prog-path``

```

jasagredo replied:
Done, thanks!

hasufell commented on doc/how-to-run-in-windows.rst:
```diff
@@ -0,0 +1,142 @@
+
+::
+
+   extra-prog-path: <msys-dir>\usr\bin
+                    <msys-dir>\<environment>\bin
```
The order is wrong. This can cause problems.

```suggestion
   extra-prog-path: <msys-dir>\<environment>\bin
                    <msys-dir>\usr\bin

```

jasagredo replied:
Done, thanks!

hasufell commented on doc/how-to-run-in-windows.rst:
```diff
@@ -0,0 +1,142 @@
+   installdir: <installdir>
+   extra-include-dirs: <msys-dir>\<environment>\include
+   extra-lib-dirs: <msys-dir>\<environment>\lib
+   extra-prog-path: <ghcup-dir>\bin
+                    <installdir>
+                    <msys-dir>\usr\bin
+                    <msys-dir>\<environment>\bin
```
```suggestion
   extra-prog-path: <ghcup-dir>\bin
                    <installdir>
                    <msys-dir>\<environment>\bin
                    <msys-dir>\usr\bin
```

jasagredo replied:
Done, thanks!

mpilgrem commented on doc/how-to-run-in-windows.rst:
```diff
@@ -0,0 +1,142 @@
+system in your computer unless told not to do so, refer to `its documentation
+<https://www.haskell.org/ghcup/install/#windows_1>`_ for more information.
+
+Ensure that Cabal can call UNIX-like tools
```
Is it accurate to categorise the tools as 'UNIX-like'? Most don't come with Windows 'out of the box', but I don't think that is unique to Windows. I would put it like this: 'Ensure that Cabal can call the tools it needs' and 'Cabal sometimes needs to call tools that do not come with Windows. The MSYS2 project makes many of them available on Windows.'

jasagredo replied:
Good point, rectified. Thanks!

Mikolaj commented:
Dear all, can we move this forward? @jasagredo: what do you need from us at this point? More comments? Reviews? Formal review acceptance? All of the above?

jasagredo commented:
I think I have addressed all the comments that arose during the reviews. I'm happy to expand this document a bit more if wanted, but I would need to know which directions should I expand it towards (debugging errors maybe?) although these could also be done in a follow-up PR.

So I think what I need is basically formal approvals, although reviews (even without approval) are also more than welcome :smile:

philderbeast commented:
> mention Stack too, even if GHCup is strongly recommended.

@mpilgrem what do you recommend? As a long time stack user, I prefer having ghcup manage the selected stack and ghc version.

```
$ cat ~/.stack/config.yaml
system-ghc: true
install-ghc: false
```

mpilgrem commented:
On 'mentioning Stack', I think it depends on the purpose of the page. If its primary purpose is the use of Cabal (the tool), provided by the `cabal-install` package, on Windows, I would not expect Stack to feature. If its primary purpose is the use of Cabal (the library) on Windows, then Stack is one way to use that library: all of Stack's building is done using Cabal (the library). Stack's User Guide has a number of references to Cabal (the tool) (I think they are all respectful - I've a 'to do' to ensure that they are accurate and up-to-date).

@philderbeast, `system-ghc` and `install-ghc` are only tangentially related to Stack using GHCup to manage versions of GHC. The better reference is: https://docs.haskellstack.org/en/stable/yaml_configuration/#ghc-installation-customisation

In terms of using GHCup (rather than Stack) to manage versions of Stack, I don't embrace that myself - but that is because I am often building Stack from source and `stack upgrade` will overwrite the GHCup 'shim' executable named `stack`.

bradrn commented on doc/how-to-run-in-windows.rst:
```diff
@@ -0,0 +1,143 @@
+`GHCup <https://www.haskell.org/ghcup/>`_. Follow the steps outlined in its
+webpage to install at least GHC and Cabal. GHCup will install its own MSYS2
+system in your computer unless told not to do so, refer to `its documentation
+<https://www.haskell.org/ghcup/install/#windows_1>`_ for more information.
```
As mentioned, I feel there should be some mention of Stack here, since it also comes with its own MSYS2 environment.

jasagredo replied:
Should this note be enough?

```rst
.. NOTE::
   Stack is another tool you can use to set up a Haskell environment. Stack
   can be installed on its own or via GHCup. See 
   `stack's webpage <https://docs.haskellstack.org/en/stable/>`_ and/or 
   `GHCup's section on stack integration <https://www.haskell.org/ghcup/guide/#stack-integration>`_,
   in particular the `Windows related subsection <https://www.haskell.org/ghcup/guide/#windows>`_.
```

bradrn replied:
I’d be happy with this. We could put GHCup in a note, too. (Maybe even a single note for both of them?)

jasagredo replied:
Perhaps I should write this as normal text instead of a note. I feel like the section `Install the Haskell environment` should cover these directly, not as side-notes.

For now it says:
- Haskell uses MSYS2
- GHCup is recommended. Install GHC and Cabal. It might bootstrap MSYS2.
- Stack also exists. It can be installed via GHCup. It might bootstrap MSYS2.

Let me rephrase the paragraph a bit and push a commit (that we can squash later on)

bradrn replied:
No, I think a note is correct, because this isn’t actually related to Cabal itself.

jasagredo replied:
I mean, how to get cabal (the executable) on Windows should be part of the Cabal docs no? 

Or do you prefer that this whole section is a note? I do think that mentioning MSYS2 and GHCup is relevant for cabal.

bradrn replied:
> I mean, how to get cabal (the executable) on Windows should be part of the Cabal docs no?

Hmm… yeah, I guess this makes sense. Whether it’s in a note or not isn’t a hugely critical point.

Also, now that I think of it, Stack is a full replacement for Cabal. If you’re using Stack, you’re probably not going to be looking at Cabal docs! So there’s probably no need to mention it at all.

jasagredo replied:
I think it depends exactly on what @mpilgrem said:

> If its primary purpose is the use of Cabal (the tool), provided by the cabal-install package, on Windows, I would not expect Stack to feature. If its primary purpose is the use of Cabal (the library) on Windows, then Stack is one way to use that library:

I personally thought this as docs for cabal-the-tool, not cabal-the-library, but as you wanted Stack to appear I am ok with mentioning it. I'm fine either way.

bradrn replied:
> I personally thought this as docs for cabal-the-tool, not cabal-the-library

Yeah, you’re right about this. I wasn’t thinking clearly when I made the suggestion. So the current text of the note is fine as-is.

jasagredo replied:
Let me push a commit right now, also including your suggestions below, and we can continue the discussion after it.

jasagredo replied:
Done, close this conversation if you are satisfied with the current contents.

Thanks!

bradrn replied:
Like I said, I’m now convinced there’s no need to mention Stack. But now that it is mentioned, I don’t strongly feel that situation needs to be changed. either. I’ve been quibbling over this documentation for long enough!

(There appears to be no close buttons for this comment…)

bradrn reviewed (approved):
This looks fine now. It’s now nicely focussed on what’s relevant to Cabal, while giving advice for other parts of the ecosystem where it can. I spot a few places where phrasing could be slightly improved, but no major issues.

hasufell commented:
> In terms of using GHCup (rather than Stack) to manage versions of Stack, I don't embrace that myself - but that is because I am often building Stack from source and stack upgrade will overwrite the GHCup 'shim' executable named stack.

Be could certainly fix that by providing a `ghcup compile stack` variant.

jasagredo commented:
> I spot a few places where phrasing could be slightly improved, but no major issues.

I'm not a native English speaker so my phrasing sometimes is not 100% correct. I will be more than happy to fix grammar/phrasing mistakes in this document should you point them out, or I happily would accept commits into this branch or a PR or something like that

bradrn commented:
> I will be more than happy to fix grammar/phrasing mistakes in this document should you point them out

OK, I’ll add some comments now.

bradrn commented on doc/how-to-run-in-windows.rst:
```diff
@@ -0,0 +1,143 @@
+The recommended way of setting up a Haskell environment in Windows is by using
+`GHCup <https://www.haskell.org/ghcup/>`_. Follow the steps outlined in its
+webpage to install at least GHC and Cabal. GHCup will install its own MSYS2
+system in your computer unless told not to do so, refer to `its documentation
```
I’d replace the comma here (before ‘refer’) with a colon. (See [run-on sentence](https://en.wikipedia.org/wiki/Sentence_clause_structure#Run-on_sentences) for details.)

jasagredo replied:
Done

bradrn commented on doc/how-to-run-in-windows.rst:
```diff
@@ -0,0 +1,143 @@
+                    <msys-dir>\usr\bin
+
+Where ``<msys-dir>`` points to the location of your MSYS2 installation. Refer to
+GHCup's documentation on where this directory is located by default.
```
Suggestion: ‘where this directory is located by default’ → ‘the default location of this directory’.

jasagredo replied:
Done

bradrn commented on doc/how-to-run-in-windows.rst:
```diff
@@ -0,0 +1,143 @@
+.. note::
+
+   Currently this step is already done by the GHCup installation on your behalf
+   by default unless told otherwise.
```
This feels a little redundant to me. I’d suggest instead: ‘Unless told otherwise, GHCup already adds these directories to `extra-prog-path` by default.’

jasagredo replied:
Done

bradrn commented on doc/how-to-run-in-windows.rst:
```diff
@@ -0,0 +1,143 @@
+.. note::
+
+   Currently this step is already done by the GHCup installation on your behalf
+   by default unless told otherwise.
```
Same here as for the previous note.

bradrn commented on doc/how-to-run-in-windows.rst:
```diff
@@ -0,0 +1,143 @@
+.. note::
+
+   Currently this step is already done by the GHCup installation on your behalf
+   by default unless told otherwise.
```
Same here as for the previous note.

jasagredo replied:
Done

bradrn commented on doc/how-to-run-in-windows.rst:
```diff
@@ -0,0 +1,143 @@
+.. note::
+
+   Currently this is already done by the GHCup installation on your behalf by
+   default unless told otherwise.
```
Same here as for the previous note.

jasagredo replied:
Done

bradrn commented:
I think it’s all fine now. GitHub says it needs two approving reviews, but after that it’s ready to be merged (in my opinion).

jasagredo commented:
I have squashed the commits into one. Now only missing piece are approving reviews :+1:

Thanks all!

Kleidukos reviewed (approved):
My thanks to @bradrn for his review of this page

hasufell commented on doc/how-to-run-in-windows.rst:
```diff
@@ -0,0 +1,153 @@
+
+.. note::
+
+   Unless told otherwise, GHCup already adds these directories to `extra-include-dirs` and `extra-lib-dirs`
+   by default.
```
The ghcup bootstrap script does. Not ghcup itself. That's an important difference.

jasagredo replied:
Ah, shall I say "When bootstrapping GHCup..."?

(side note, I see you use `ghcup` not `GHCup`. Do you have a preference that I change mentions to `ghcup` instead?)

hasufell replied:
'GHCup' is the project, `ghcup` usually refers to the binary. The bootstrap script bootstraps the binary. On windows, it also installs msys2 (the powershell script) and changes the cabal config (the bash script, which is invoked by the powershell script and is the same that you run on linux) as you pointed out. The info doesn't need to be that detailed, but should refer to the bootstrap scripts.

jasagredo replied:
Done :+1:

ffaf1 reviewed (commented):
pretty fantastic

Mikolaj commented:
Shall we backport this to 3.12, since this only expands (and plausibly fixes) documentation? I assume it does not describe anything that is on master branch, but not the 3.12 branch (which is master from 1 or 2 months ago)?

jasagredo commented:
These are purely additions to the documentation, and the behavior described has not changed for a long time, so it is applicable to 3.12 (even to older versions). So if possible, backporting it would be nice.

Mikolaj commented:
@mergify backport 3.12
